src/raftstore/store/fsm/peer.rs

```diff
@@ -1086,54 +1120,58 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
                         new_region_id, r, new_region
                     );
                 }
+                self.ctx.router.stop(new_region_id);
             }
 
-            let mut new_peer = match Peer::create(self, &new_region) {
-                Ok(new_peer) => new_peer,
+            let (sender, mut new_peer) = match PeerFsm::create(self.ctx, &new_region) {
+                Ok((sender, new_peer)) => (sender, new_peer),
                 Err(e) => {
                     // peer information is already written into db, can't recover.
                     // there is probably a bug.
                     panic!("create new split region {:?} err {:?}", new_region, e);
                 }
             };
-            let peer = new_peer.peer.clone();
+            let meta_peer = new_peer.peer.peer.clone();
 
-            for peer in new_region.get_peers() {
+            for p in new_region.get_peers() {
                 // Add this peer to cache.
-                new_peer.insert_peer_cache(peer.clone());
+                new_peer.peer.insert_peer_cache(p.clone());
             }
 
             // New peer derive write flow from parent region,
             // this will be used by balance write flow.
-            new_peer.peer_stat = peer_stat.clone();
-
-            let campaigned = new_peer.maybe_campaign(is_leader, &mut self.pending_raft_groups);
+            new_peer.peer.peer_stat = self.fsm.peer.peer_stat.clone();
+            let campaigned = new_peer.peer.maybe_campaign(is_leader);
+            new_peer.has_ready |= campaigned;
 
             if is_leader {
                 // The new peer is likely to become leader, send a heartbeat immediately to reduce
                 // client query miss.
-                new_peer.heartbeat_pd(&self.pd_worker);
+                new_peer.peer.heartbeat_pd(&self.ctx.pd_scheduler);
             }
 
-            new_peer.activate();
-            self.region_peers.insert(new_region_id, new_peer);
+            new_peer.peer.activate(self.ctx);
+            meta.regions.insert(new_region_id, new_region);
+            if last_region_id == new_region_id {
+                // To prevent from big region, the right region needs run split
```

Comment:
already take the approximate_size above, so here no need to set `siez_diff_hint` anymore

Reply:
> Can we remove this logic since we already using batch split?

I don't think batch split can prevent big region.

> already take the approximate_size above, so here no need to set siez_diff_hint anymore

Let's keep it the same old way. And you can make the optimization in other pull request.